keps/sig-api-machinery/20190325-unions.md

```diff
@@ -0,0 +1,213 @@
+will be interpreted as an intention to clear all the other fields. See section
+below.
+
+### Normalizing on updates
+
+A "normalization process" will run automatically for all creations and
+modifications (either with update or patch). It will happen automatically in order
+to clear fields and update discriminators. This process will run for both
+core-types and CRDs. It will take place before validation. The sent object
+doesn't have to be valid, but fields will be cleared in order to make it valid.
+This process will also happen before fields tracking (server-side apply), so
+changes in discriminator, even if implicit, will be owned by the client making
+the update (and may result in conflicts).
+
+This process works as follows:
+- If there is a discriminator, and its value has changed, clear all fields but
+  the one specified by the discriminator,
+- If there is no discriminator, or if its value hasn't changed,
+  - if there is exactly one field, set the discriminator when there is one
+    to that value. Otherwise,
+  - compare the fields set before and after. If there is exactly one field
+    added, set the discriminator (if present) to that value, and remove all
+    other fields. if more than one field has been added, leave the process so
+    that validation will fail.
+
+#### "At most one" versus "exactly one"
+
+The goal of this proposal is not to change the validation, but to help client
+clear other fields in the union. Validation should be implemented for in-tree
+types as it is today, or through "oneOf" properties in CRDs.
+
+In other word, this is proposing to implement "at most one", and the exactly one
+should be provided through another layer of validation (separated).
+
+#### Clearing all the fields
+
+Since the system is trying to do the right thing, it can be hard to "clear
+everything". In that case, each API could decide to have their own "Nothing"
+value in the discriminator, which will automatically trigger a clearing of all
+fields beside "Nothing".
+
+### Backward compatibilities properties
+
+This normalization process has a few nice properties, especially for dumb
+clients, when it comes to backward compatibility:
+- A dumb client that doesn't know which fields belong to the union can just set
+  a new field and get all the others cleared automatically
+- A dumb client that doesn't know about the discriminator is going to change a
+  field, leave the discriminator as it is, and should still expect the fields to
+  be cleared accordingly
+- A dumb client that knows about the discriminator can change the discriminator
+  without knowing which fields to clear, they will get cleared automatically
+
+
+### Validation
+
+Objects have to be validated AFTER the normalization process, which is going to
+leave multiple fields of the union if it can't normalize. As discussed in
+drawbacks below, it can also be useful to validate apply requests before
+applying them.
```

Comment:
would normalization be part of the defaulter?

Reply:
No, for server-side apply it happens BEFORE we default, for all other modifying end-points, it happens after defaulting. It's part of the "field management"

Reply:
ok, so the plumbing is mostly in place. good.